README.rst
================================
  MDAnalysis Repository README
================================

|build| |cov| [*]_

|docs| |devdocs| |usergroup| |developergroup| |anaconda| |mybinder|


here is text for a test commit

MDAnalysis_ is a Python toolkit to analyze molecular dynamics
trajectories generated by a wide range of popular simulation packages
including DL_Poly, CHARMM, Amber, NAMD, LAMMPS, and Gromacs. (See the
lists of  supported `trajectory formats`_ and `topology formats`_.)

.. code:: python

   import MDAnalysis as mda

   # Load simulation results with a single line
   u = mda.Universe('topol.tpr','traj.trr')

   # Select atoms
   ag = u.select_atoms('name OH')

   # Atom data made available as Numpy arrays
   ag.positions
   ag.velocities
   ag.forces

   # Iterate through trajectories
   for ts in u.trajectory:
       print(ag.center_of_mass())

There are also a number of tutorials_ on the MDAnalysis homepage that explain
how to conduct RMSD calculations, Alignment and more features of MDAnalysis.

Source code
===========

Source code is hosted in a git repository at

https://github.com/MDAnalysis/mdanalysis

and is available under the GNU General Public License, version 2 (see
the file LICENSE_).

This is the top level of the master repository. It contains

1. the MDAnalysis toolkit source files in the directory ::

      package/

2. the unit tests together with any input files required for
   running those tests in the directory ::

      testsuite/

The directory ``maintainer`` contains scripts only needed for
maintaining releases and are not generally useful for the user or the
typical developer.

(For more details on the directory layout see `Issue 87`_ on the
MDAnalysis issue tracker.)

Guide for Developers
====================

To setup a development environment and run the testsuite you can use this
guide_. If you are a new developer who would like to start contributing to
MDAnalysis as a start you can increase our code coverage, the guides explain how
to find uncovered code.

.. Footnotes

.. [*] **build**: Unit testing is for the whole package; **coverage** is
       shown for the core library modules and the analysis modules.

.. _trajectory formats: https://docs.mdanalysis.org/documentation_pages/coordinates/init.html#id1
.. _topology formats: https://docs.mdanalysis.org/documentation_pages/topology/init.html#supported-topology-formats
.. _Issue 87: https://github.com/MDAnalysis/mdanalysis/issues/87
.. _MDAnalysis: https://www.mdanalysis.org
.. _LICENSE: https://github.com/MDAnalysis/mdanalysis/blob/master/LICENSE
.. _`#286`: https://github.com/MDAnalysis/mdanalysis/issues/286
.. _`MDAnalysis.analysis`: https://docs.mdanalysis.org/documentation_pages/analysis_modules.html
.. _`tutorials`: https://www.mdanalysis.org/pages/learning_MDAnalysis/
.. _`guide`: https://github.com/MDAnalysis/mdanalysis/wiki/Guide-for-Developers

.. |usergroup| image:: https://img.shields.io/badge/Google%20Group-Users-lightgrey.svg
   :alt: User Google Group
   :target: http://users.mdanalysis.org

.. |developergroup| image:: https://img.shields.io/badge/Google%20Group-Developers-lightgrey.svg
   :alt: Developer Google Group
   :target: http://developers.mdanalysis.org

.. |docs| image:: https://img.shields.io/badge/docs-latest-brightgreen.svg
   :alt: Documentation (latest release)
   :target: https://docs.mdanalysis.org

.. |devdocs| image:: https://img.shields.io/badge/docs-development-yellow.svg
   :alt: Documentation (development version)
   :target: https://www.mdanalysis.org/mdanalysis/

.. |build| image:: https://travis-ci.org/MDAnalysis/mdanalysis.svg?branch=develop
   :alt: Build Status
   :target: https://travis-ci.org/MDAnalysis/mdanalysis

.. |cov|   image:: https://codecov.io/gh/MDAnalysis/mdanalysis/branch/develop/graph/badge.svg
   :alt: Coverage Status
   :target: https://codecov.io/gh/MDAnalysis/mdanalysis

.. |anaconda| image:: https://anaconda.org/conda-forge/mdanalysis/badges/version.svg
   :alt: Anaconda
   :target: https://anaconda.org/conda-forge/mdanalysis

.. |mybinder| image:: https://mybinder.org/badge.svg
   :alt: My Binder
   :target: https://mybinder.org/v2/gh/MDAnalysis/binder-notebook/master


package/CHANGELOG
# -*- tab-width: 2; indent-tabs-mode: nil; coding: utf-8 -*-
======================
 MDAnalysis CHANGELOG
======================

The rules for this file:
  * entries are sorted newest-first.
  * summarize sets of changes - don't reproduce every git log comment here.
  * don't ever delete anything.
  * keep the format consistent (79 char width, M/D/Y date format) and do not
    use tabs but use spaces for formatting
  * accompany each entry with github issue/PR number (Issue #xyz)
  * release numbers follow "Semantic Versioning" http://semver.org

------------------------------------------------------------------------------
mm/dd/yy mattwthompson, richardjgowers
  * 0.20.0

Enhancements
  * Added analysis.polymer.sort_backbone to rearrange backbone atoms into
    sequential order (bondwise)

Fixes
  * fixed lib.nsgrid segfault for coordinates very close to the box boundary
    (Issue #2132, PR #2136)

Deprecations
  * analysis.polymer.PersistenceLength no longer requires the perform_fit()
    method to be used, this is done automatically in run()

11/06/18 richardjgowers
  * 0.19.2

Changes
  * Added Windows support


11/05/18 orbeckst, PicoCentauri, richardjgowers, zemanj
  * 0.19.1

Fixes
   * limit output of Chainreader __repr__ (#2109)
   * added missing docs for lib.pkdtree (#2104)
   * Added sphinx markup for FrameIterator (#2106)
   * fixed error for residues and segments containing 0 atoms (#1999)
   * fixed numpy deprecation warnings (#2122, #2123)
   * Fixed reading PDB files with DOS line ending (#2128)

Deprecations
   * Default ``filename`` directory of align.AlignTraj is deprecated and 
     will change in 1.0 to the current directory.

10/09/18 tylerjereddy, richardjgowers, palnabarun, orbeckst, kain88-de, zemanj,
         VOD555, davidercruz, jbarnoud, ayushsuhane, hfmull, micaela-matta,
         sebastien.buchoux, arm61, p-j-smith, IAlibay

  * 0.19.0

Enhancements
  * Added bond/angle/dihedral reading in PARM7 TOPParser (PR #2052)
  * Replaced multiple apply (_apply_distmat, _apply_kdtree)
    methods in distance based selections with
    lib.distances.capped_distance for automatic selection of
    distance evaluation method. (PR #2035)
  * Modified analysis.rdf.InterRDF to use lib.distances.capped_distance
    to handle rdf calculations for large systems. (PR #2013)
  * Added return_distances argument in lib.distances.capped_distances
    to evaluate and return distances only when required. Modified
    the optimization rules in lib.distances._determine_method for
    faster computations. (PR #2041)
  * Added method search_tree in lib.pkdtree to find all the pairs between
    two kdtrees. (PR #2041)
  * Added a wrapper of lib.nsgrid in lib.distances.self_capped_distance
    and lib.distances.capped_distanceto automatically chose the fastest
    method for distance based calculations. (PR #2008)
  * Added Grid search functionality in lib.nsgrid for faster distance based
    calculations. (PR #2008)
  * Modified around selections to work with KDTree and periodic boundary
    conditions.  Should reduce memory usage (#974 PR #2022)
  * Modified topology.guessers.guess_bonds to automatically select the
    fastest method for guessing bonds using
    lib.distance.self_capped_distance (PR # 2006)
  * Added lib.distances.self_capped_distance to internally select the
    optimized method for distance evaluations of coordinates with itself. (PR # 2006)
  * Added augment functionality to create relevant images of particles
    in the vicinity of central cell to handle periodic boundary
    conditions (PR #1977)
  * Added lib.distances.capped_distance function to quickly calculate all distances
    up to a given maximum distance (PR #1941)
  * Added a rotation coordinate transformation (PR #1937)
  * Added a box centering trajectory transformation (PR #1946)
  * Added a on-the-fly trajectory transformations API and a coordinate translation
    function (Issue #786)
  * Added various duecredit stubs
  * Import time reduced by a factor two (PR #1881)
  * Added compound kwarg to center, centroid, center_of_geometry, center_of_mass (PR #1903)
  * Added rdf.InterRDF_s to calculate site-specific pair distribution
    functions (PR #1815)
  * Increased performance of (repeated) calls to AtomGroup.pack_into_box()
    (PR #1922)
  * Added boolean property *Group.isunique (PR #1922)
  * Added *Group.copy() methods returning an identical copy of the respective
    group (PR #1922)
  * Use a faster function to deduplicate indices (PR #1951)
  * Calculations in *Group.center() are performed in double precision (#PR1936)
  * Functions in lib.distances accept coordinate arrays of arbitrary dtype
    (PR #1936)
  * Added pbc kwarg to Bond/Angle/Dihedral/Improper object value method,
    default True. (Issue #1938)
  * ChainReader can correctly handle continuous trajectories split into multiple files,
    generated with gromacs -noappend (PR #1728)
  * MDAnalysis.lib.mdamath now supports triclinic boxes and rewrote in Cython (PR #1965)
  * AtomGroup.write can write a trajectory of selected frames (Issue #1037)
  * Added dihedrals.py with Dihedral, Ramachandran, and Janin classes to
    analysis module (PR #1997, PR #2033)
  * Added the analysis.data module for reference data used in analysis (PR #2033)
  * Added analysis.dihedrals with Ramachandran class to analysis module (PR #1997)
  * Added augment functionality to create relevant images of particles
  * Most functions in `MDanalysis.lib.distances` previously only accepting
    arrays of coordinates now also accept single coordinates as input (PR #2048,
    Issues #1262 #1938)
  * Performance improvements to make_whole (PR #1965)
  * Performance improvements to fragment finding (PR #2028)
  * Added user-defined boxes in density code (PR #2005)
  * MemoryReader now can accept velocities and forces (PR #2080)
  * Universe.transfer_to_memory now copies dimensions, velocities and forces
    (where possible) (Issue #1041 PR #2080)

Fixes

testsuite/MDAnalysisTests/analysis/test_persistencelength.py
# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding:utf-8 -*-
# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4 fileencoding=utf-8
#
# MDAnalysis --- https://www.mdanalysis.org
# Copyright (c) 2006-2017 The MDAnalysis Development Team and contributors
# (see the file AUTHORS for the full list of names)
#
# Released under the GNU Public Licence, v2 or any higher version
#
# Please cite your use of MDAnalysis in published work:
#
# R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler,
# D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein.
# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
# doi: 10.25080/majora-629e541a-00e
#
# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
#
from __future__ import print_function, division, absolute_import

import pytest

import MDAnalysis as mda
from MDAnalysis.analysis import polymer
from MDAnalysis.exceptions import NoDataError

import numpy as np
import matplotlib

from numpy.testing import assert_almost_equal, assert_equal

from MDAnalysisTests.datafiles import Plength, TRZ_psf, TRZ


class TestPersistenceLength(object):
    @staticmethod
    @pytest.fixture()
    def u():
        return mda.Universe(Plength)

    @staticmethod
    @pytest.fixture()
    def p(u):
        ags = [r.atoms.select_atoms('name C* N*')
               for r in u.residues]

        p = polymer.PersistenceLength(ags)
        return p

    @staticmethod
    @pytest.fixture()
    def p_run(p):
        return p.run()

    def test_ag_ValueError(self, u):
        ags = [u.atoms[:10], u.atoms[10:110]]
        with pytest.raises(ValueError):
            polymer.PersistenceLength(ags)

    def test_run(self, p_run):
        assert len(p_run.results) == 280

    def test_lb(self, p_run):
        assert_almost_equal(p_run.lb, 1.485, 3)

    def test_fit(self, p_run):
        assert_almost_equal(p_run.lp, 6.504, 3)
        assert len(p_run.fit) == len(p_run.results)

    def test_plot_ax_return(self, p_run):
        '''Ensure that a matplotlib axis object is
        returned when plot() is called.'''
        actual = p_run.plot()
        expected = matplotlib.axes.Axes
        assert isinstance(actual, expected)

    def test_perform_fit_warn(self, p_run):
        with pytest.warns(DeprecationWarning):
            p_run.perform_fit()

class TestFitExponential(object):
    x = np.linspace(0, 250, 251)
    a_ref = 20.0
    y = np.exp(-x / a_ref)

    def test_fit_simple(self):
        a = polymer.fit_exponential_decay(self.x, self.y)
        assert a == self.a_ref

    def test_fit_noisy(self):
        noise = np.sin(self.x) * 0.01
        y2 = noise + self.y

        a = polymer.fit_exponential_decay(self.x, y2)

        assert_almost_equal(a, self.a_ref, decimal=3)
        # assert np.rint(a) == self.a_ref

class TestSortBackbone(object):
    @staticmethod
    @pytest.fixture(scope='class')
    def u():
        return mda.Universe(TRZ_psf, TRZ)

    def test_missing_bonds(self):
        u = mda.Universe(Plength)

        with pytest.raises(NoDataError):
            polymer.sort_backbone(u.atoms[:10])

    def test_sortbb(self, u):
        # grab backbone atoms out of order
        # 0 1 4 6 8 - correct
        ag = u.atoms[[4, 1, 0, 8, 6]]

        s_ag = polymer.sort_backbone(ag)

        assert_equal(s_ag.ids, [0, 1, 4, 6, 8])

    def test_not_fragment(self, u):
        # two fragments don't work
        bad_ag = u.residues[0].atoms[:2] + u.residues[1].atoms[:2]
        with pytest.raises(ValueError):
            polymer.sort_backbone(bad_ag)

    def test_branches(self, u):
        # includes side branches, can't sort
        bad_ag = u.atoms[:10]  # include -H etc

        with pytest.raises(ValueError):
            polymer.sort_backbone(bad_ag)


package/MDAnalysis/analysis/polymer.py
# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding:utf-8 -*-
# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
#
# MDAnalysis --- https://www.mdanalysis.org
# Copyright (c) 2006-2017 The MDAnalysis Development Team and contributors
# (see the file AUTHORS for the full list of names)
#
# Released under the GNU Public Licence, v2 or any higher version
#
# Please cite your use of MDAnalysis in published work:
#
# R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler,
# D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein.
# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
# doi: 10.25080/majora-629e541a-00e
#
# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
#

"""
Polymer analysis --- :mod:`MDAnalysis.analysis.polymer`
=======================================================


:Author: Richard J. Gowers
:Year: 2015, 2018
:Copyright: GNU Public License v3

This module contains various commonly used tools in analysing polymers.


Example
-------

Finding the persistence length of a polymer:

>>> from MDAnalysis.tests.datafiles import TRZ_psf, TRZ
>>> import MDAnalysis as mda
>>> from MDAnalysis.analysis import polymer
>>> u = mda.Universe(TRZ_psf, TRZ)
>>> # this system is a pure polymer melt, so we can select
>>> # the chains by using the .fragments attribute
>>> chains = u.atoms.fragments
>>> # sort the chains, removing any none backbone atoms
>>> sorted_chains = [polymer.sort_backbone(f.select_atoms('not name O* H'))
...                  for f in chains]
>>> lp = polymer.PersistenceLength(sorted_chains)
>>> lp = lp.run()
>>> print('The persistence length is: {}'.format(lp.pl))
>>> # always check the visualisation of this:
>>> lp.plot()

"""
from __future__ import division, absolute_import
from six.moves import range

import numpy as np
import scipy.optimize
import warnings
import logging

from .. import NoDataError
from ..core.groups import requires, AtomGroup
from ..lib.distances import calc_bonds
from .base import AnalysisBase

logger = logging.getLogger(__name__)


@requires('bonds')
def sort_backbone(backbone):
    """Rearrange a linear AtomGroup into backbone order

    Requires that the backbone has bond information,
    and that only backbone atoms are provided (ie no side
    chains or hydrogens).

    Parameters
    ----------
    backbone : AtomGroup
      the backbone atoms, not necessarily in order

    Returns
    -------
    sorted_backbone : AtomGroup
      backbone in order, so `sorted_backbone[i]` is bonded to
      `sorted_backbone[i - 1]` and `sorted_backbone[i + 1]`


    .. versionadded:: 0.20.0
    """
    if not len(backbone.fragments) == 1:
        raise ValueError("Multiple fragments found in backbone.  "
                         "backbone must be a single contiguous AtomGroup")
    if max(len(at.bonded_atoms & backbone) for at in backbone) > 2:
        # find which atom has too many bonds for easier debug
        branches = [at for at in backbone
                    if len(at.bonded_atoms & backbone) > 2]
        raise ValueError(
            "backbone is not linear.  "
            "The following atoms have more than two bonds in backbone: "
            "{}".format(','.join(str(a) for a in branches)))

    caps = [atom for atom in backbone
           if len(atom.bonded_atoms & backbone) == 1]

     # choose one of the capping hydrogens to be the startpoint
    sorted_backbone = AtomGroup([caps[0]])

    # iterate until the sorted chain length matches the backbone size
    while len(sorted_backbone) < len(backbone):
        # current end of the chain
        end_atom = sorted_backbone[-1]

        # look at all bonded atoms which are also part of the backbone
        # and subtract any that have already been added
        next_atom = (end_atom.bonded_atoms & backbone) - sorted_backbone

        # append this to the sorted backbone
        sorted_backbone += next_atom

    return sorted_backbone


class PersistenceLength(AnalysisBase):
    r"""Calculate the persistence length for polymer chains

    The persistence length is the length at which two points on the polymer
    chain become decorrelated.  This is determined by first measuring the
    autocorrelation (:math:`C(n)`) of two bond vectors (:math:`\mathbf{a}`)
    separated by :math:`n` bonds

    .. math::

       C(n) = \langle \cos\theta_{i, i+n} \rangle =
               \langle \mathbf{a_i} \cdot \mathbf{a_{i+n}} \rangle

    An exponential decay is then fitted to this, which yields the
    persistence length

    .. math::

       C(n) \approx \exp\left( - \frac{n l_B}{l_P} \right)

    where :math:`l_B` is the average bond length, and :math:`l_P` is the
    persistence length.

    Parameters
    ----------
    atomgroups : list
       List of AtomGroups.  Each should represent a single
       polymer chain, ordered in the correct order.
    verbose : bool (optional)
       Show detailed progress of the calculation if set to ``True``; the
       default is ``False``.

    Attributes
    ----------
    results : numpy.ndarray
       the measured bond autocorrelation
    lb : float
       the average bond length
    lp : float
       calculated persistence length
    fit : numpy.ndarray
       the modelled backbone decorrelation predicted by *lp*

    See Also
    --------
    :func:`sort_backbone`
       for producing the sorted AtomGroup required for input.

    .. versionadded:: 0.13.0
    .. versionchanged:: 0.20.0
       The run method now automatically performs the exponential fit
    """
    def __init__(self, atomgroups, **kwargs):
        super(PersistenceLength, self).__init__(
            atomgroups[0].universe.trajectory, **kwargs)
        self._atomgroups = atomgroups

        # Check that all chains are the same length
        lens = [len(ag) for ag in atomgroups]
        chainlength = len(atomgroups[0])
        if not all(l == chainlength for l in lens):
            raise ValueError("Not all AtomGroups were the same size")

        self._results = np.zeros(chainlength - 1, dtype=np.float32)

    def _single_frame(self):
        # could optimise this by writing a "self dot array"
        # we're only using the upper triangle of np.inner
        # function would accept a bunch of coordinates and spit out the
        # decorrel for that
        n = len(self._atomgroups[0])

        for chain in self._atomgroups:
            # Vector from each atom to next
            vecs = chain.positions[1:] - chain.positions[:-1]
            # Normalised to unit vectors
            vecs /= np.sqrt((vecs * vecs).sum(axis=1))[:, None]

            inner_pr = np.inner(vecs, vecs)
            for i in range(n-1):
                self._results[:(n-1)-i] += inner_pr[i, i:]

    def _conclude(self):
        n = len(self._atomgroups[0])

        norm = np.linspace(n - 1, 1, n - 1)
        norm *= len(self._atomgroups) * self.n_frames

        self.results = self._results / norm
        self._calc_bond_length()

        self._perform_fit()

    def _calc_bond_length(self):
        """calculate average bond length"""
        bs = []
        for ag in self._atomgroups:
            pos = ag.positions
            b = calc_bonds(pos[:-1], pos[1:]).mean()
            bs.append(b)
        self.lb = np.mean(bs)

    def perform_fit(self):
        warnings.warn("perform_fit is now called automatically from run",
                      DeprecationWarning)

    def _perform_fit(self):
        """Fit the results to an exponential decay"""
        try:
            self.results
        except AttributeError:
            raise NoDataError("Use the run method first")
        self.x = np.arange(len(self.results)) * self.lb

        self.lp = fit_exponential_decay(self.x, self.results)

        self.fit = np.exp(-self.x/self.lp)

    def plot(self, ax=None):
        """Visualise the results and fit

        Parameters
        ----------
        ax : matplotlib.Axes, optional
          if provided, the graph is plotted on this axis

        Returns
        -------
        ax : the axis that the graph was plotted on
        """
        import matplotlib.pyplot as plt
        if ax is None:
            ax = plt.gca()
        ax.plot(self.x, self.results, 'ro', label='Result')
        ax.plot(self.x, self.fit, label='Fit')
        ax.set_ylabel(r'$C(x)$')

        ax.set(xlabel='x', ylabel='C(x)', xlim=[0.0, 40 * self.lb])
        ax.legend(loc='best')

        return ax


def fit_exponential_decay(x, y):
    r"""Fit a function to an exponential decay

    .. math::  y = \exp(- \frac{x}{a})

    Parameters
    ----------
    x, y : array_like
      The two arrays of data

    Returns
    -------
    a : float
      The coefficient *a* for this decay

    Notes
    -----
    This function assumes that data starts at 1.0 and decays to 0.0

    """
    def expfunc(x, a):
        return np.exp(-x/a)

    a = scipy.optimize.curve_fit(expfunc, x, y)[0][0]

    return a
